parity/main.rs

```diff
@@ -323,26 +324,28 @@ impl Configuration {
 		// Sync
 		let sync = EthSync::register(service.network(), sync_config, client);
 
-		// Setup rpc
-		if self.args.flag_jsonrpc {
-			setup_rpc_server(service.client(), sync.clone(), &self.args.flag_jsonrpc_url, &self.args.flag_jsonrpc_cors);
-			SocketAddr::from_str(&self.args.flag_jsonrpc_url).unwrap_or_else(|_|die!("{}: Invalid JSONRPC listen address given with --jsonrpc-url. Should be of the form 'IP:port'.", self.args.flag_jsonrpc_url));
-		}
-
 		// Register IO handler
 		let io_handler  = Arc::new(ClientIoHandler {
 			client: service.client(),
 			info: Default::default(),
-			sync: sync
+			sync: sync.clone(),
 		});
 		service.io().register_handler(io_handler).expect("Error registering IO handler");
 
+		// Setup rpc
+		let server_handler = if self.args.flag_jsonrpc {
+			SocketAddr::from_str(&self.args.flag_jsonrpc_url).unwrap_or_else(|_|die!("{}: Invalid JSONRPC listen address given with --jsonrpc-url. Should be of the form 'IP:port'.", self.args.flag_jsonrpc_url));
+			setup_rpc_server(service.client(), sync, &self.args.flag_jsonrpc_url, &self.args.flag_jsonrpc_cors)
+		} else {
+			None
+		};
+
 		// Handle exit
-		wait_for_exit(&service);
+		wait_for_exit(&service, server_handler);
 	}
 }
 
-fn wait_for_exit(client_service: &ClientService) {
+fn wait_for_exit(client_service: &ClientService, server_handler: Option<Arc<PanicHandler>>) {
```

Comment:
Instead of having two `MayPanic` (one of them as `Option<Arc<PanicHandler>>`) you could create just single `PanicHandler` and use `forward_from` on client_service and rpc_server (or `PanicHandler`).